File: Array/zhanchengguo/0701-返回倒数第k个节点/Main.java
package com.test.ktapplication;


import android.util.Log;

public class Main {

    void main() {
        ListNode a = new ListNode(1);
        ListNode b = new ListNode(2);
        ListNode c = new ListNode(3);
        ListNode d = new ListNode(4);
        ListNode e = new ListNode(5);
        a.next = b;
        b.next = c;
        c.next = d;
        d.next = e;
        e.next = null;
        int result = kthToLast(a, 2);
        Log.e("value is ", result + "");
    }

    public int kthToLast(ListNode head, int k) {
        ListNode fast = head;
        ListNode slow = head;
        while (k-- > 0) {
            fast = fast.next;
        }
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        return slow.val;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

File: Array/zhanchengguo/0701-返回倒数第k个节点/main.kt
package com.test.ktapplication

fun main() {
    val a = ListNode(1)
    val b = ListNode(2)
    val c = ListNode(3)
    val d = ListNode(4)
    val e = ListNode(5)
    a.next = b
    b.next = c
    c.next = d
    d.next = e
    e.next = null
    var result = kthToLast(a, 3)
    print("value is ${result}\n")
}

fun kthToLast(head: ListNode?, k: Int): Int? {
    var fast = head
    var slow = head
    var num = k
    while (num-- > 0) {
        fast = fast?.next
    }
    while (fast != null) {
        fast = fast?.next
        slow = slow?.next
    }
    return slow?.`val`
}


data class ListNode(var `val`: Int) {
    var next: ListNode? = null
}